### PR TITLE
add ci cd with circleci

### PR DESCRIPTION
**Pull request by naesheim**
The declarative pipeline is found in the .circleci folder.

I have also done some changes on the source code:
- I changed the S3 bucket name in order to fit my aws.
- I changed the hardcoded tmp/ destination for policy. Because the folder structure of circleci demanded so.

For this to work you need to:
- enable this project in circle ci.
- add aws credentials in project settings.
- change s3 name.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      - checkout
+      - run:
+          name: pack python folder
+          command: |
+            python3 -m venv venv
+            venv/bin/pip3 install -r requirements.txt
+            cd venv/lib/python3.6/site-packages
+            zip -r $OLDPWD/ImageResize.zip .
+            cd $OLDPWD
+            zip ImageResize.zip image_resize.py
```
Comment (naesheim) [inline .circleci/config.yml]: Makes a zipped folder with dependencies and the python file

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      
+workflows:
+  version: 2
+
+  build-test-deploy:
+    jobs:
+      - build:
+          filters:
+            branches:
+              only:
```
Comment (naesheim) [inline .circleci/config.yml]: If this is removed - this will run on master aswell

File: image_resize.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+        print("bucket: " + str(bucket))
+        print("key: "+str(key))
+
+        s3_client.download_file(bucket, key, original_img)
+
+        for size_px in THUMBNAIL_SIZES_PX:
+            resize_image(original_img, temp_img, size_px)
+            new_name_key = key.replace(".jpg", "_"+str(size_px)+".jpg")
+            s3_client.upload_file(temp_img, '{}-resized'.format(bucket), new_name_key )
```
Comment (cankarm) [inline image_resize.py]: This approach is strange to me. It seems that you would like to add this file (which is a part of different repo) in current repo. I think that this should be downloaded only to your circleci environment and there used _sed_ to substitute parameters where needed.

File: playbooks/aws_role/create.yml [hunk 4]
```diff
@@ -6,7 +6,7 @@
         name: "{{ role_name }}"
         state: present
         description: "{{ role_description }}"
-        assume_role_policy_document: "{{ lookup('file','/tmp/policy.json') }}"
+        assume_role_policy_document: "{{ lookup('file','/tmp/resources/policy.json') }}"
```
Comment (cankarm) [inline playbooks/aws_role/create.yml:9]: Same as before. I do not see the reason to change this in the current repo just so the circleci step will be easier.

File: playbooks/lambda/create.yml [hunk 5]
```diff
@@ -30,6 +30,5 @@
 
     - name: Set attributes
       set_stats:
-        data:
+        data: 
           function_arn: "{{alias_info.alias_arn}}"
-
```
Comment (cankarm) [inline playbooks/lambda/create.yml:35]: Is this change necessary?

File: requirements.txt [hunk 6]
```diff
@@ -0,0 +1,2 @@
+boto3
+Pillow==6.1.0
```
Comment (cankarm) [inline requirements.txt]: Similar as before, copy from (an)other repository.

File: resize_service_opera.yml [hunk 7]
```diff
@@ -140,7 +140,7 @@ topology_template:
       properties:
         aws_region: eu-central-1
         function_name: radon-toy-example-image-res
-        zip_file: /tmp/X-test-ImageRes.zip
+        zip_file: /tmp/resources/ImageResize.zip
         function_alias: test
         permission_id: lambda_test_permission01
```
Comment (cankarm) [inline resize_service_opera.yml]: Changes that should be intrinsic to circleci step.

File: resize_service_opera.yml [hunk 8]
```diff
@@ -170,9 +170,4 @@ topology_template:
       requirements:
         - host: my-workstation
         - receives_notification: aws_lambda
-        - bucket: bucket_in
-
-
-
-
-
+        - bucket: bucket_in
```
Comment (cankarm) [inline resize_service_opera.yml]: Is this necessary? At least one newline at the end of the file is nice to have.

File: resize_service_opera.yml [hunk 9]
```diff
@@ -150,15 +150,15 @@ topology_template:
...
+        bucket_name: radon-thumbnail
         aws_region: eu-central-1
       requirements:
         - host: my-workstation
     
     bucket_out:
       type: radon.nodes.s3_bucket
       properties:
-        bucket_name: radon-img3-resized
+        bucket_name: radon-thumbnail-resized
```
Comment (cankarm) [inline resize_service_opera.yml]: Changes that should be intrinsic to the circleci step.

**Review by cankarm — CHANGES_REQUESTED** [review]
This PR is OK from the CircleCI configuration approach but I think that it should not add files from "image resize" repo nor it should not make file changes in this current repo. To change some configurations, CircleCI step should change configuration lines (names of the function, buckets) only "internaly" and not on the public repo.

File: image_resize.py (hunk 3 above)
Comment (naesheim) [reply image_resize.py]: Yes. Maybe that is a whole different discussion. How to include the actual source code. Maybe its here the Function Hub comes into the picture?

File: playbooks/aws_role/create.yml (hunk 4 above)
Comment (naesheim) [reply playbooks/aws_role/create.yml:9]: Fair enough! I can redo that.

File: resize_service_opera.yml (hunk 9 above)
Comment (naesheim) [reply resize_service_opera.yml]: @cankarm  does xopera allow for overwrite properties?

File: playbooks/lambda/create.yml (hunk 5 above)
Comment (naesheim) [reply playbooks/lambda/create.yml:35]: No:)

File: requirements.txt (hunk 6 above)
Comment (naesheim) [reply requirements.txt]: The idea with ci/cd is to link the source code to the deployment. 

The most common way is to let them share repository. The other is to have a hardcoded, versioned artifact you pull in at test/deploy stage.

File: resize_service_opera.yml (hunk 7 above)
Comment (naesheim) [reply resize_service_opera.yml]: Yep:)

File: resize_service_opera.yml (hunk 8 above)
Comment (naesheim) [reply resize_service_opera.yml]: Not intentionally.

File: image_resize.py (hunk 3 above)
Comment (cankarm) [reply image_resize.py]: Interesting thought. In the sense of RADON and my opinion, the orchestrator will get CSAR from the IDE or from the CI/CD tool, which would take the same version that IDE creates, but alter it with some configs. Like you did.

Currently I see the *yaml file as one big config. Sure it should be improved. But as this is a kind of config, I would understand that CI should not modify the file in repo, but just alter it before passing it to the orchestrator. Is this possible? Should we use some other think as "submodule"?
